### DISCUSSION TITLE
Provide a Method to Preprocess Gherkin Documents Before Test Class Generation

### DISCUSSION THREAD

**Discussion by serge-rsvz-inasti** [reqnroll/Reqnroll#192] [Ideas]
Hello,

Which of the following two examples do you prefer?

```
Examples:
    | animal_a          | animal_b      | a_is_stronger_than_b |
    | {cat, dog, lion } | {mouse, worm} | true                 |
    | {mouse, worm, cat}| {dog, lion }  | false                |
```

```
Examples:
    | animal_a | animal_b | a_is_stronger_than_b |
    | cat      | mouse    | true                 |
    | cat      | worm     | true                 |
    | dog      | mouse    | true                 |
    | dog      | worm     | true                 |
    | lion     | mouse    | true                 |
    | lion     | worm     | true                 |
    | mouse    | dog      | false                |
    | mouse    | lion     | false                |
    | worm     | dog      | false                |
    | worm     | lion     | false                |
    | cat      | dog      | false                |
    | cat      | lion     | false                |
```

For the project I'm working on, we felt having a comfortable framework was paramount for test adoption. One limitation we faced with the vanilla Gherkin transformation to test class was the inability to express data with multiple representations. For instance, expressing sets and ranges required each possible representation case to be written in full, making test authoring tedious.

I checked if a plugin could handle this, but it currently can't. I then created a custom tool to preprocess raw Gherkin files to complete vanilla ones, but this approach is not as pleasing as having the process handled within a plugin. So, calling a method that could be injected through plugins to preprocess the Gherkin would be awesome.

**Comment by gasparnagy** [comment]
I think the external data plugin (https://github.com/reqnroll/Reqnroll/tree/main/Plugins/Reqnroll.ExternalData/Reqnroll.ExternalData.ReqnrollPlugin) does such pre-processing. Check the [ExternalDataTestGenerator](https://github.com/reqnroll/Reqnroll/blob/main/Plugins/Reqnroll.ExternalData/Reqnroll.ExternalData.ReqnrollPlugin/ExternalDataTestGenerator.cs) class. I think based on that you should be able to make a plugin that preprocesses your example as well.

**Reply by serge-rsvz-inasti** [reply]
To my understanding, the implementation of that plugin is closely intertwined with core components, making it difficult to reproduce outside of the reqnroll solution. This limits the possibility of creating a plugin that can coexist with others.

**Reply by gasparnagy** [reply]
I don't think so. Originally that plugin was done fully external. Now it is part of the main repo, but it only uses the API that is usable to everyone (at least as far as I am aware). Do you have any concrete detail in that plugin that you puzzle with?
